enterprise/server/usage/usage.go

```diff
@@ -0,0 +1,138 @@
+	// to the DB.
+	//
+	// NOTE: if this value is changed, then the implementation of
+	// collectionPeriodStartUsec needs to change as well. Also note that if this
+	// value is increased, we may lose some usage data for a short period of time,
+	// since keys at smaller granularity may not be properly flushed to the DB.
+	collectionPeriodDuration = 1 * time.Minute
+
+	// redisKeyTTL defines how long usage keys have to live before they are
+	// deleted automatically by Redis.
+	//
+	// Keys should live for at least 2 collection periods since collection periods
+	// aren't finalized until the period is past, plus some wiggle room for Redis
+	// latency. We add a few more collection periods on top of that, in case
+	// flushing fails due to transient errors.
+	redisKeyTTL = 5 * collectionPeriodDuration
+
+	redisUsageKeyPrefix  = "usage/"
+	redisGroupsKeyPrefix = redisUsageKeyPrefix + "groups/"
+	redisCountsKeyPrefix = redisUsageKeyPrefix + "counts/"
+)
+
+type TrackerOpts struct {
+	Clock timeutil.Clock
+}
+
+type tracker struct {
+	env   environment.Env
+	rdb   *redis.Client
+	clock timeutil.Clock
+}
+
+func NewTracker(env environment.Env, opts *TrackerOpts) (*tracker, error) {
+	rdb := env.GetCacheRedisClient()
+	if rdb == nil {
+		return nil, status.UnimplementedError("Missing redis client for usage")
+	}
+	clock := opts.Clock
+	if clock == nil {
+		clock = timeutil.NewClock()
+	}
+	return &tracker{
+		env:   env,
+		rdb:   rdb,
+		clock: clock,
+	}, nil
+}
+
+func (ut *tracker) Increment(ctx context.Context, counts *tables.UsageCounts) error {
+	groupID, err := perms.AuthenticatedGroupID(ctx, ut.env)
+	if err != nil {
+		return err
+	}
+	if groupID == "" {
+		// Don't track anonymous usage for now.
+		return nil
+	}
+
+	m, err := countsToMap(counts)
+	if err != nil {
+		return err
+	}
+	if len(m) == 0 {
+		return nil
+	}
+
+	t := ut.clock.Now()
+
+	pipe := ut.rdb.TxPipeline()
+	// Add the group ID to the set of groups with usage
+	groupsKey := groupsRedisKey(t)
+	pipe.SAdd(ctx, groupsKey, groupID)
+	pipe.Expire(ctx, groupsKey, redisKeyTTL)
+	// Increment the hash values
+	countsKey := countsRedisKey(groupID, t)
+	for k, c := range m {
+		pipe.HIncrBy(ctx, countsKey, k, c)
+	}
+	pipe.Expire(ctx, countsKey, redisKeyTTL)
+	if _, err := pipe.Exec(ctx); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func collectionPeriodStartUsec(t time.Time) int64 {
+	utc := t.UTC()
+	// Start of minute
+	start := time.Date(
+		utc.Year(), utc.Month(), utc.Day(),
+		utc.Hour(), utc.Minute(), 0, 0,
+		utc.Location())
+	return timeutil.ToUsec(start)
+}
+
+func groupsRedisKey(t time.Time) string {
+	return fmt.Sprintf("%s%d", redisGroupsKeyPrefix, collectionPeriodStartUsec(t))
+}
+
+func countsRedisKey(groupID string, t time.Time) string {
+	return fmt.Sprintf("%s%s/%d", redisCountsKeyPrefix, groupID, collectionPeriodStartUsec(t))
+}
+
+func countsToMap(tu *tables.UsageCounts) (map[string]int64, error) {
+	b, err := json.Marshal(tu)
```

Comment:
What do you think about explicitly mapping the fields instead of relying on this json back & forth conversion?
Renaming a field in the struct will cause problems for usage tracking & there will be no direct code references to individual fields to suggest it's a bad idea.

Reply:
Sounds good, made the names `underscore_case` to simplify the DB field mapping in the next PR